tests/blackbox_asserts.py

```diff
@@ -0,0 +1,298 @@
+                        DryRunEncoder.hex(math.gcd(*args)) if math.gcd(*args) else None
+                    ),
+                    DRProp.finalScratch: lambda args, actual: is_subdict(
+                        {0: math.gcd(*args)}, actual
+                    ),
+                    DRProp.stackTop: lambda args: math.gcd(*args),
+                    DRProp.maxStackHeight: 2,
+                    DRProp.status: lambda args: "PASS" if math.gcd(*args) else "REJECT",
+                    DRProp.passed: lambda args: bool(math.gcd(*args)),
+                    DRProp.rejected: lambda args: not bool(math.gcd(*args)),
+                    DRProp.errorMessage: None,
+                },
+            }
+
+            exec_mode = mode_to_execution_mode(Mode.Application)
+            inputs, assertions = SequenceAssertion.inputs_and_assertions(scenario, exec_mode)
+            euclid_results = Executor.dryrun_app_on_sequence(algod, euclid_app_teal, inputs)
+            for assert_type, predicate in assertions.items():
+                assertion = SequenceAssertion(predicate, name=str(assert_type))
+                assertion.dryrun_assert(inputs, euclid_results, assert_type)
+    """
+    input_types = subr.subroutine.input_types
+    assert (
+        input_types is not None
+    ), "please provide input_types in your @Subroutine annotation (crucial for generating proper end-to-end testable PyTeal)"
+
+    arg_names = subr.subroutine.arguments()
+
+    def arg_prep_n_call(i, p):
+        name = arg_names[i]
+        by_ref = name in subr.subroutine.by_ref_args
+        arg_expr = Txn.application_args[i] if mode == Mode.Application else Arg(i)
+        if p == TealType.uint64:
+            arg_expr = Btoi(arg_expr)
+        prep = None
+        arg_var = arg_expr
+        if by_ref:
+            arg_var = ScratchVar(p)
+            prep = arg_var.store(arg_expr)
+        return prep, arg_var
+
+    def subr_caller():
+        preps_n_calls = [*(arg_prep_n_call(i, p) for i, p in enumerate(input_types))]
+        preps, calls = zip(*preps_n_calls) if preps_n_calls else ([], [])
+        preps = [p for p in preps if p]
+        invocation = subr(*calls)
+        if preps:
+            return Seq(*(preps + [invocation]))
+        return invocation
+
+    def make_return(e):
+        if e.type_of() == TealType.uint64:
+            return e
+        if e.type_of() == TealType.bytes:
+            return Len(e)
+        if e.type_of() == TealType.anytype:
+            x = ScratchVar(TealType.anytype)
+            return Seq(x.store(e), Int(1337))
+        # TealType.anytype:
+        return Seq(e, Int(1337))
+
+    def make_log(e):
+        if e.type_of() == TealType.uint64:
+            return Log(Itob(e))
+        if e.type_of() == TealType.bytes:
+            return Log(e)
+        return Log(Bytes("nada"))
+
+    if mode == Mode.Signature:
+
+        def approval():
+            return make_return(subr_caller())
```

Comment:
Shouldn't this also have a case for the return type being `TealType.none`?

Reply:
This case - `returnType == TealType.none` for `mode.Signature` is covered by [this unit test](https://github.com/algorand/pyteal/blob/005f72e671eecd82cc151179b051d2a370704d53/tests/blackbox_test.py#L137)

Also the semantic tests for [@Subroutine square_byref()](https://github.com/algorand/pyteal/blob/005f72e671eecd82cc151179b051d2a370704d53/tests/blackbox_test.py#L91) which are running in PySDK prove that it works as expected.